mesa/space.py

```diff
@@ -485,7 +486,439 @@ def exists_empty_cells(self) -> bool:
+
+    def _get_neighborhood_mask(
+        self, pos: Coordinate, moore: bool, include_center: bool, radius: int
+    ) -> np.ndarray:
+        """
+        Generate a boolean mask representing the neighborhood.
+        Helper method for select_cells_multi_properties() and move_agent_to_random_cell()
+
+        Args:
+            pos (Coordinate): Center of the neighborhood.
+            moore (bool): True for Moore neighborhood, False for Von Neumann.
+            include_center (bool): Include the central cell in the neighborhood.
+            radius (int): The radius of the neighborhood.
+
+        Returns:
+            np.ndarray: A boolean mask representing the neighborhood.
+        """
+        neighborhood = self.get_neighborhood(pos, moore, include_center, radius)
+        mask = np.zeros((self.width, self.height), dtype=bool)
+
+        # Convert the neighborhood list to a NumPy array and use advanced indexing
+        coords = np.array(neighborhood)
+        mask[coords[:, 0], coords[:, 1]] = True
+        return mask
+
+    def select_cells_multi_properties(
+        self,
+        conditions: dict,
+        only_neighborhood: bool = False,
+        pos: None | Coordinate = None,
+        moore: bool = True,
+        include_center: bool = False,
+        radius: int = 1,
+    ) -> list[Coordinate]:
+        """
+        Select cells based on multiple property conditions using NumPy, optionally within a neighborhood.
+
+        Args:
+            conditions (dict): A dictionary where keys are property names and values are
+                               callables that take a single argument (the property value)
+                               and return a boolean.
+            only_neighborhood (bool): If True, restrict selection to the neighborhood.
+            pos, moore, include_center, radius: Optional neighborhood parameters.
+
+        Returns:
+            List[Coordinate]: Coordinates where conditions are satisfied.
+        """
+        # Start with a mask of all True values
+        combined_mask = np.ones((self.width, self.height), dtype=bool)
+
+        for prop_name, condition in conditions.items():
+            prop_layer = self.properties[prop_name].data
+            # Apply the condition to the property layer
+            prop_mask = condition(prop_layer)
+            # Combine with the existing mask using logical AND
+            combined_mask = np.logical_and(combined_mask, prop_mask)
+
+        if only_neighborhood and pos is not None:
+            neighborhood_mask = self._get_neighborhood_mask(
+                pos, moore, include_center, radius
+            )
+            combined_mask = np.logical_and(combined_mask, neighborhood_mask)
+
+        # Extract coordinates from the combined mask
+        selected_cells = list(zip(*np.where(combined_mask)))
+        return selected_cells
+
+    def move_agent_to_random_cell(
+        self,
+        agent: Agent,
+        conditions: dict,
```

Comment:
cool idea to randomly move, but only to a specified neighborhood.

Reply:
Thanks! I think generalizing that is indeed a good idea, I replied below.

Reply:
Done, you can now insert any mask. Utility methods `get_neighborhood_mask()` and `get_empty_mask` are provided.